spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java

```diff
@@ -163,7 +164,10 @@ else if (StringUtils.hasText(asyncExecutor)) {
 
 		gatewayProxyBuilder.addConstructorArgValue(serviceInterface);
 
-		return new BeanDefinitionHolder(gatewayProxyBuilder.getBeanDefinition(), id);
+		AbstractBeanDefinition beanDefinition = gatewayProxyBuilder.getBeanDefinition();
+		beanDefinition.addMetadataAttribute(new BeanMetadataAttribute("factoryBeanObjectType", serviceInterface));
```

Comment:
Note that this solution won't work if the service interface is a property placeholder (rare, perhaps, but possible).

We may just have to live with that, or boot will have to resolve property placeholders in `BeanTypeRegistry.determineTypeFromDefinitionAttribute()`.